lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,574 @@
+                    }
+                }
+                return false;
+                ;
+              })
+              .Case<RegOp, RegResetOp>([&](auto) {
+                children.clear();
+                return true;
+              })
+              .Case<SubfieldOp, SubindexOp, SubaccessOp>([&](auto sub) {
+                // If the subfield result is not recorded then it must be
+                // ignored. Only aggregate port and wire ops are recorded.
+                if (valToStorageLocIdMap.find(node) ==
+                    valToStorageLocIdMap.end()) {
+                  children.clear();
+                  return true;
+                }
+                return false;
+                ;
+              })
+              .Default([&](auto) { return false; });
+      if (skip)
+        return;
+    }
+
+    // Now, handle the ops which create an explicit comb path from `node` to
+    // another Value.
+    for (auto &child : node.getUses()) {
+      Operation *owner = child.getOwner();
+      LLVM_DEBUG(llvm::dbgs() << "\n owner: " << *owner);
+      auto childVal =
+          TypeSwitch<Operation *, Value>(owner)
+              .Case<FConnectLike>([&](FConnectLike connect) {
+                // Ignore connects, where the child is the destination.
+                if (child.getOperandNumber() == 0)
+                  return (Value) nullptr;
+                // Get the aggregate base type Value, if the dest is a subfield
+                // of it. Returns an invalid FieldRef, if this Value can be
+                // ignored.
+                auto fRef = getFieldRef(connect.getDest());
+
+                if (fRef.getValue() == nullptr)
+                  return (Value) nullptr;
+                if (inputArg.isValid())
+                  if (auto sinkOut = fRef.getValue().dyn_cast<BlockArgument>())
+                    // Record the combinational path that was discovered between
+                    // the ports of this module.
+                    portPaths[inputArg.val][inputArg.fieldId].push_back(
+                        Node(sinkOut, fRef.getFieldID()));
+
+                return connect.getDest();
+              })
+              .Case<RegOp, RegResetOp, SubaccessOp>(
+                  [&](auto reg) { return (Value) nullptr; })
+              .Default([&](Operation *op) {
+                if (op->getNumResults() == 1) {
+                  return (Value)op->getResult(0);
+                }
+                return (Value) nullptr;
+              });
+      if (childVal != nullptr) {
+        children.push_back({childVal, owner});
+      }
+    }
+  }
+
+  /// Record all the leaf ground type values for an aggregate type value. The
+  /// Optional `fieldIdsToAdd` can be used to only add field ids that exist in
+  /// the set and ignore all other fields. This is used for MemOp ports.
+  void gatherAggregateLeafs(Value val,
+                            llvm::SmallDenseSet<size_t> fieldIdsToAdd = {}) {
```

Comment:
Can we avoid copy?

Reply:
The set would just have 3 elements (Memory ports), when present, so I think copy shouldn't be an issue ?

Reply:
That makes sense!